### ISSUE TITLE
Calling a keras model on its inputs leads to slower training

### ISSUE THREAD

**Issue by wateryzephyr** [keras-team/keras#13155]
**System information**  
- Have I written custom code: **yes**
- OS Platform and Distribution: **Linux Ubuntu 16.04**
- TensorFlow backend:  **yes**
- TensorFlow version: **tensorflow-gpu 1.13.1** 
- Keras version: **2.2.4**
- Python version:  **2.7**
- CUDA/cuDNN version: **CUDA 10, CUDNN 7**
- GPU model and memory: **Titan Xp, 12GB**

**Describe the current behavior**  
When a keras model that has layer updates (e.g. batch norm moving averages and variances) is called with the inputs it was originally defined with, the training becomes significantly slower:

```python
model = keras.models.Model(inputs=inputs, outputs=outputs)
pred = model(inputs)  # Slow, compared to using model.outputs
```

**Describe the expected behavior**  
I would expect keras model's input / output tensor caching mechanism to prevent the updates being added twice, therefore leading to no differences in speed.

**Code to reproduce the issue**  
I've created a small repo to repro the issue [here](https://github.com/wateryzephyr/keras_bn_bug_repro). A docker file is included to help weed out environmental / version issues.

**Other info / logs**  

Timings with `model.outputs`:
```
Step 0: 1.046618
Took 4.6513
Step 100: 3.962632
Took 6.9885
Step 200: 2.113625
Took 6.9877
Step 300: 1.654034
Took 6.9938
Step 400: 1.457640
Took 6.9861
Step 500: 1.351904
Took 6.9923
Step 600: 1.285378
Took 6.9826
Step 700: 1.228966
Took 6.9860
Step 800: 1.202688
Took 6.9880
Step 900: 1.178856
Took 6.9878
```

Timings with `model(inputs)`:
```
Step 0: 1.044049
Took 5.2239
Step 100: 2.616739
Took 8.1010
Step 200: 1.510669
Took 8.0969
Step 300: 1.276879
Took 8.0946
Step 400: 1.191719
Took 8.0972
Step 500: 1.143867
Took 8.0973
Step 600: 1.111212
Took 8.0994
Step 700: 1.098528
Took 8.1017
Step 800: 1.076799
Took 8.1016
Step 900: 1.068770
Took 8.0945
```

This is ~10-15% slower, and is exacerbated in multi-GPU settings (~40% in some applications).

**Comment by wateryzephyr** [comment]
Some more info: one WAR I tried was

```diff
diff --git a/keras/engine/network.py b/keras/engine/network.py
index 530ab9d..8da6b3d 100644
--- a/keras/engine/network.py
+++ b/keras/engine/network.py
@@ -247,6 +247,12 @@ class Network(Layer):
              input_shapes=[x._keras_shape for x in self.inputs],
              output_shapes=[x._keras_shape for x in self.outputs])
 
+        # Update the cache.
+        cache_key = object_list_uid(self.inputs)
+        cache_key += '_' + object_list_uid([None] * len(self.inputs))
+        if cache_key not in self._output_tensor_cache:
+            self._output_tensor_cache[cache_key] = self.outputs
+
         # Fill in the output mask cache.
         masks = []
         for x in self.inputs:
```

This effectively resolved the slowdown; however, this changes the behavior where tensors from a model would have `model_name` pre-pended to `layer_name/output_tensor_name`. Any advice on a cleaner fix would be much appreciated.

**Comment by wateryzephyr** [comment]
@gowtham-kp Any advice on how to fill up the `_output_tensor_cache` with outputs that are properly named?

**Comment by wateryzephyr** [comment]
@gowthamkpr Any updates on this issue?
